contributors/design-proposals/defer-containers.md

```diff
@@ -0,0 +1,248 @@
+semantics. The proposed DeferContainers will provide destructor semantics for such stateful PODs
+
+## Abstract
+This introduces the concept of ‘deferContainers’, inspired by golang’s keyword ‘defer’. This will allow one to define a
+set of Containers to be executed at the time of POD termination. The defined containers will be executed sequentially and
+in the specified order. It will bring destructor() capability to Pods.
+
+## Use Cases
+
+### Cleanup:
+Most stateful workloads require certain Cleanup activity before or after the appContainers are terminated such as
+* Sync/Flush the Disk or logs before the POD is evicted off a node.
+* Delete or Update Global Configuration / Release Application level Lock.
+* Some Legacy applications may not respond to SIGTERM signals sent by docker daemon, they might require special command/procedure
+  to initiate graceful shutdown.
+
+### Shard rebalancing.
+In a sharded workload configured as a statefulset, if the application is scaled down, then the exiting Pod should trigger
+a re-balance of keys from the current shard to the rest of them. This needs to be done befoe the shard goes down.  
+‘deferContainers’ could handle such scenarios with ease.
+
+### Statefulset Upgrading
+If in future, statefulsets support rolling updates, an update request might attempt to replace Pods one after the other. 
+With deferContainers each Pod could be removed gracefully without having to implement complicated sidecar logic to prevent data loss.
+
+### Master-Slave or Leader-Follower Statefulset down-size
+Consider a Stateful set consisting a Master-Slave or Leader-follower type application running N replicas. Ordinal index 
+ranging from 0 to N-1.  If you scale it down (reduce the number of replicas), the statefulset controller would attempt 
+to bring down the last replica (N-1th pod in this case).  If that pod is the elected Leader or Master, then this would 
+disrupt the service intermittently. With deferContainers, this handover or re-election could be programmed gracefully/elegantly.
+
+### SHUTDOWN Sequence
+* Traditional relational  databases typically support a reasonable shutdown sequence, for instance, Oracle has 4 types of 
+shutdown such as NORMAL, IMMEDIATE, TRANSACTIONAL and ABORT. ‘deferContainers’ will allow us to program and wait for such
+complex shutdown scenarios. 
+* In future when kuberenetes supports Virtual Machine runtime (eg: hyperContainer) for better isolation, we should shutdown
+  the VMs instead of killing them abruptly.  ‘deferContainers’ could help us run such shutdwon commands.
+
+## Limitation with the current system
+Container pre-stop hooks are not sufficient for all termination cases:
+* They cannot easily coordinate complex conditions across containers in multi-container pods
+* They can only function with code in the image or code in a shared volume, which would have to be statically linked 
```

Comment:
The objective is not always to pull a new image, the objective is to seperate out/de-couple Termination sequence. Most often we wont have to pull a new image all togather.  I can also imagine simple cleanups like 'rm -rf /tmp/dir1' wrapped around bash or busyboxy images which are lighter.

If your pod is part of a CI engine and running 1000s of testcases, it should be easy to configure a deferContainer (from a different image/source) simply upload (curl POST) currently processed test results to a different server when the POD is getting evicted out of that node.  
If curl is not available in your primary container you dont have to re-package and create a new image just to run curl at the termination - which will be the case for preStop hooks.